src/circuit/gadget/ecc.rs

```diff
@@ -12,129 +12,227 @@ use halo2::{
 
     /// Performs variable-base scalar multiplication, returning `[scalar] base`.
     fn mul(
-        layouter: &mut impl Layouter<Self>,
-        scalar: &Self::Scalar,
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        scalar: C::Scalar,
         base: &Self::Point,
     ) -> Result<Self::Point, Error>;
 
-    /// Performs fixed-base scalar multiplication, returning `[scalar] base`.
+    /// Performs fixed-base scalar multiplication using a full-width scalar, returning `[scalar] base`.
     fn mul_fixed(
-        layouter: &mut impl Layouter<Self>,
-        scalar: &Self::Scalar,
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        scalar: &Self::ScalarFixed,
+        base: &Self::FixedPoint,
+    ) -> Result<Self::Point, Error>;
+
+    /// Performs fixed-base scalar multiplication using a short signed scalar, returning `[scalar] base`.
+    fn mul_fixed_short(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        scalar: &Self::ScalarFixedShort,
         base: &Self::FixedPoint,
     ) -> Result<Self::Point, Error>;
 }
 
-/// An element of the given elliptic curve's scalar field.
+/// A full-width element of the given elliptic curve's scalar field, to be used for fixed-base scalar mul.
 #[derive(Debug)]
-pub struct Scalar<C: CurveAffine, EccChip: EccInstructions<C>> {
-    inner: EccChip::Scalar,
+pub struct ScalarFixed<C: CurveAffine, EccChip: EccInstructions<C> + Clone> {
+    chip: EccChip,
+    inner: EccChip::ScalarFixed,
 }
 
-impl<C: CurveAffine, EccChip: EccInstructions<C>> Scalar<C, EccChip> {
-    /// Constructs a new point with the given value.
+impl<C: CurveAffine, EccChip: EccInstructions<C> + Clone> ScalarFixed<C, EccChip> {
+    /// Constructs a new ScalarFixed with the given value.
+    pub fn new(
+        chip: EccChip,
+        mut layouter: impl Layouter<C::Base>,
+        value: Option<C::Scalar>,
+    ) -> Result<Self, Error> {
+        chip.witness_scalar_fixed(&mut layouter, value)
+            .map(|inner| ScalarFixed {
+                chip: chip.clone(),
+                inner,
+            })
+    }
+}
+
+/// A signed short element of the given elliptic curve's scalar field, to be used for fixed-base scalar mul.
+#[derive(Debug)]
+pub struct ScalarFixedShort<C: CurveAffine, EccChip: EccInstructions<C> + Clone> {
+    chip: EccChip,
+    inner: EccChip::ScalarFixedShort,
+}
+
+impl<C: CurveAffine, EccChip: EccInstructions<C> + Clone> ScalarFixedShort<C, EccChip> {
+    /// Constructs a new ScalarFixedShort with the given value.
     pub fn new(
-        mut layouter: impl Layouter<EccChip>,
+        chip: EccChip,
+        mut layouter: impl Layouter<C::Base>,
         value: Option<C::Scalar>,
     ) -> Result<Self, Error> {
-        EccChip::witness_scalar(&mut layouter, value).map(|inner| Scalar { inner })
+        chip.witness_scalar_fixed_short(&mut layouter, value)
+            .map(|inner| ScalarFixedShort {
+                chip: chip.clone(),
```

Comment:
No need to clone here, since `chip` is passed by value (implying the caller has already cloned it if necessary).
```suggestion
                chip,
```